src/api/resources/index.ts

```diff
@@ -0,0 +1,519 @@
+          if (!validate.chartId(apiData.id)) {
+            res.status(406).send(`Invalid chart resource id supplied!`)
+            return
+          }
+        } else {
+          if (!validate.uuid(apiData.id)) {
+            res.status(406).send(`Invalid resource id supplied!`)
+            return
+          }
+        }
+
+        try {
+          const retVal = await this.resProvider[apiData.type]?.setResource(
+            apiData.type,
+            apiData.id,
+            apiData.value
+          )
+          this.server.handleMessage(
+            this.resProvider[apiData.type]?.pluginId as string,
+            this.buildDeltaMsg(apiData.type, apiData.id, apiData.value)
+          )
+          res.status(200).send(`SUCCESS: ${req.params.resourceType} complete.`)
+        } catch (err) {
+          res.status(404).send(`ERROR: ${req.params.resourceType} incomplete!`)
+        }
+      }
+    )
+  }
+
+  private processApiRequest(req: Request) {
+    let resType: SignalKResourceType = 'waypoints'
+    if (req.params.resourceType.toLowerCase() === 'waypoint') {
+      resType = 'waypoints'
+    }
+    if (req.params.resourceType.toLowerCase() === 'route') {
+      resType = 'routes'
+    }
+    if (req.params.resourceType.toLowerCase() === 'note') {
+      resType = 'notes'
+    }
+    if (req.params.resourceType.toLowerCase() === 'region') {
+      resType = 'regions'
+    }
+    if (req.params.resourceType.toLowerCase() === 'charts') {
+      resType = 'charts'
+    }
+
+    const resValue: any = buildResource(resType, req.body)
+
+    const resId: string = req.params.resourceId
+      ? req.params.resourceId
+      : (resType = 'charts' ? resValue.identifier : UUID_PREFIX + uuidv4())
+
+    return {
+      type: resType,
+      id: resId,
+      value: resValue
+    }
+  }
+
+  private getResourcePaths(): { [key: string]: any } {
+    const resPaths: { [key: string]: any } = {}
+    for (const i in this.resProvider) {
+      if (this.resProvider.hasOwnProperty(i)) {
+        resPaths[i] = {
+          description: `Path containing ${
+            i.slice(-1) === 's' ? i.slice(0, i.length - 1) : i
+          } resources`,
+          $source: this.resProvider[i]?.pluginId
+        }
+      }
+    }
+    return resPaths
+  }
+
+  private checkForProvider(resType: SignalKResourceType): boolean {
+    debug(`** checkForProvider(${resType})`)
+    debug(this.resProvider[resType])
+
+    if (this.resProvider[resType]) {
+      if (
+        !this.resProvider[resType]?.listResources ||
+        !this.resProvider[resType]?.getResource ||
+        !this.resProvider[resType]?.setResource ||
+        !this.resProvider[resType]?.deleteResource ||
+        typeof this.resProvider[resType]?.listResources !== 'function' ||
+        typeof this.resProvider[resType]?.getResource !== 'function' ||
+        typeof this.resProvider[resType]?.setResource !== 'function' ||
+        typeof this.resProvider[resType]?.deleteResource !== 'function'
```

Comment:
This code block checks that the resProvider has all the four ResourceProviderMethods, right? Shouldn't we check this when the plugin calls `register`, so we can fail the registration and log an error, to get an early warning? Then we can skip the runtime check in all http method handlers and just assume that if there is a registered handler we can just call it, no `checkForProvider` call needed?

Reply:
I have moved the ResourceProviderMethods check to `register()` but the `checkForProvider()` remains as it is used by the route handlers to determine whether to pass a request to a provider or call `next()`